File: 739.daily-temperatures.java
/*
 * @lc app=leetcode id=739 lang=java
 *
 * [739] Daily Temperatures
 */

// @lc code=start
class Solution {
    public int[] dailyTemperatures(int[] temperatures) {
        Stack<Integer> stack = new Stack<>();
        int n = temperatures.length;
        int[] ans = new int[n];
        
        for(int i=n-1; i>=0; --i) {
            while(!stack.empty() && temperatures[stack.peek()] <= temperatures[i]) {
                stack.pop();
            }
            ans[i] = stack.empty()? 0 : stack.peek() - i;
            stack.push(i);
        }
        return ans;
    }
}
// @lc code=end

File: solutions/739.daily-temperatures.rb
# @param {Integer[]} temperatures
# @return {Integer[]}
def daily_temperatures(temperatures)
  n = temperatures.size
  st = []
  ans = []
  (n - 1).downto(0).each do |i|
    t = temperatures[i]
    st.pop while !st.empty? && temperatures[st.last] <= t
    ans[i] = st.empty? ? 0 : st.last - i
    st << i
  end
  ans
end
